design/src/rfcs/rfc0027_improve_sdk_request_id_access.md

```diff
@@ -0,0 +1,244 @@
+   which has a `request_id` accessor on it.
+4. For success cases, the customer can't retrieve the request ID at all if they use the fluent
+   client. Instead, they must manually make the operation and call the underlying Smithy client
+   so that they have access to `SdkSuccess`, which provides the raw response where the request ID
+   can be manually extracted from headers.
+
+Only one of these mechanisms is convenient and ergonomic. The rest need considerable improvements.
+Additionally, the request ID should be attached to tracing events where possible so that enabling
+debug logging reveals the request IDs without any code changes being necessary.
+
+This RFC proposes changes to make the request ID easier to access.
+
+Terminology
+-----------
+
+- **Request ID:** A unique identifier assigned to and associated with a request to AWS that is
+  sent back in the response headers. This identifier is useful to customers when requesting support.
+- **Operation Error:** Operation errors are code generated for each operation in a Smithy model.
+  They are an enum of every possible modeled error that that operation can respond with, as well
+  as an `Unhandled` variant for any unmodeled or unrecognized errors.
+- **Modeled Errors:** Any error that is represented in a Smithy model with the `@error` trait.
+- **Unmodeled Errors:** Errors that a service responds with that do not appear in the Smithy model.
+- **SDK Clients:** Clients generated for the AWS SDK, including "adhoc" or "one-off" clients.
+- **Smithy Clients:** Any clients not generated for the AWS SDK, excluding "adhoc" or "one-off" clients.
+
+SDK/Smithy Purity
+-----------------
+
+Before proposing any changes, the topic of purity needs to be covered. Request IDs are not
+currently a Smithy concept. However, at time of writing, the request ID concept is leaked into
+the non-SDK rust runtime crates and generated code via the [generic error] struct and the
+`request_id` functions on generated operation errors (e.g., [`GetObjectError` example in S3]).
+
+This RFC attempts to remove these leaks from Smithy clients.
+
+Proposed Changes
+----------------
+
+First, we'll explore making it easier to retrieve a request ID from errors,
+and then look at making it possible to retrieve them from successful responses.
+To see the customer experience of these changes, see the **Example Interactions**
+section below.
+
+### Make request ID retrieval on errors consistent
+
+One could argue that customers being able to convert a `SdkError` into an operation error
+that has a request ID on it is sufficient. However, there's no way to write a function
+that takes an error from any operation and logs a request ID, so it's still not ideal.
+
+The `aws-http` crate needs to have a `RequestId` trait on it to facilitate generic
+request ID retrieval:
+
+```rust
+pub trait RequestId {
+    /// Returns the request ID if it's available.
+    fn request_id(&self) -> Option<&str>;
+}
```

Comment:
I like this UX. From a user's perspective, I think it's what I'd try to call before even reading the docs. I think that we one thing that could possibly be expanded upon is that we may need to educate users on when to expect this to return `Some` vs. `None`. In some cases, returning `None` is expected but users may find code paths that produce an unexpected `None` and we need to direct them to file a bug report for that.